### ISSUE TITLE
'contains' operator not working ok with other conditions

### ISSUE THREAD

**Issue by hmgheorghe** [orientechnologies/orientdb#3526]
Hi guys,

apparently contains works well when alone in where clause but has issues when paired with other conditions.

select from concept where synonym contains 'cell line'
returned 1 result

select from concept where synonym contains 'cell line' and name='cell_line'
returned 0 results

![screen shot 2015-02-04 at 16 53 32](https://cloud.githubusercontent.com/assets/10850420/6042551/7d24f9de-ac8e-11e4-8e57-87326bc595c0.png)
![screen shot 2015-02-04 at 16 53 53](https://cloud.githubusercontent.com/assets/10850420/6042550/7d189e8c-ac8e-11e4-8e4a-80c3973e535a.png)

**Comment by vitorenesduarte** [comment]
Hi,
```sql
create class concept
create property concept.name string
create property concept.synonym embeddedlist string
insert into concept set name = 'cell_line', synonym = ['Cell Line', 'cell line']
select from concept where synonym contains 'cell line' and name = 'cell_line'
```
@hmgheorghe does this work for you?

**Comment by lvca** [comment]
The issue has been closed because no response has been received after a while. In case the issue is still valid, please reopen it or write a comment that explains why it's still valid and a way to reproduce it. Thank you.

**Comment by sowee15** [comment]
Hi,

I couldn't reproduce the issue with the particular example above but I can with type DATE or DATETIME. I created the following use case:

````
create class concept_dates
create property concept_dates.name string
create property concept_dates.dates_set embeddedset date
create property concept_dates.dates_list embeddedlist date
create property concept_dates.datetimes_set embeddedset datetime
create property concept_dates.datetimes_list embeddedlist datetime
insert into concept_dates set name = 'mydates', dates_set = ['2015-10-19','2015-10-20'], dates_list = ['2015-10-19','2015-10-20'], datetimes_set = ['2015-10-19 00:00:00','2015-10-20 00:00:00'], datetimes_list = ['2015-10-19 00:00:00','2015-10-20 00:00:00']
````

Then all those SELECT return 0 result.

````
select from concept_dates where dates_set contains "2015-10-20"
select from concept_dates where dates_set contains "2015-10-20 00:00:00"
select from concept_dates where dates_list contains "2015-10-20"
select from concept_dates where dates_list contains "2015-10-20 00:00:00"
select from concept_dates where datetimes_list contains "2015-10-20 00:00:00"
select from concept_dates where datetimes_set contains "2015-10-20 00:00:00"
````

I then created a NOTUNIQUE SB_TREE index on field dates_set and ran the second SELECT above again and then I had one result.

I deleted the index, ran the query again, no result

I recreated the index, ran the query again, 1 result. So far its consistent, but I don't understand why the index is needed in the first place.

Then I added a second condition:

````
select from concept_dates where dates_set contains "2015-10-20" and name = "mydates"
````

Got 1 result.

I created a new NOTUNIQUE SBTREE index on field name then ran the previous query again. I got no result. Is this expected? At this point, if I just run the query with the contains but without the condition on the name, I still get 1  result.

I tried creating a composite NOTUNIQUE SBTREE index on both dates_set and name fields, I still got 0 result when running the query with both conditions.

What's weird is that I made the same test with integers instead of dates and the contains work the first time, without ever needing an index.

Please let me know what you think of my tests, if its normal behavior or a bug. Thanks!

I run OrientDB 2.1.4 on windows 8.1. All my queries have been made in the web studio.

**Comment by luigidellaquila** [comment]
Hi @sowee15 

ok, I'm reopening it.

Thanks for the detailed test case

Luigi

**Comment by luigidellaquila** [comment]
Hi guys, 

I just fixed it in develop branch, now I'm backporting the fix to 2.1.x

Luigi

**Comment by sowee15** [comment]
Can I download a compiled fixed version to test it?

**Comment by luigidellaquila** [comment]
HI @sowee15 
in next hours it will be available as a snapshot from here:
https://oss.sonatype.org/content/repositories/snapshots/com/orientechnologies/orientdb-community/

As an alternative, you can build it from sources:
- clone git repository locally
- switch to 2.1.x branch (git checkout 2.1.x):
- execute the following command from the main folder of OrientDB
. _base/script/deploy.sh 2.1.5-SNAPSHOT
It will build OrientDB distribution and it will change the current directory right where the distro has been created.

**Comment by sowee15** [comment]
Thanks! I'll check it out.

**Comment by sowee15** [comment]
I'm currently building it on my windows laptop using cygwin, maven 3.3.3
and jdk 1.8.0.45, once its done should I have something stable or I'm
better off waiting for the snapshot on the other site you sent me?

Thanks!

**Comment by luigidellaquila** [comment]
Well, it depends on what you mean with "stable", it's a snapshot ;-)
Anyway, it's enough stable to check if the issue is solved, then you can wait for next official hotfix release, that is typically released weekly

Thanks

Luigi

**Comment by sowee15** [comment]
I meant could I run on discrepancies because I built on windows. Got it for
the fact that it's a snapshot. The build was successful, I'll test it out
and report back and then will look out for the next hotfix. What's the best
way to know when a hotfix is released? So far we've just been sticking to
production releases.

Thanks!

**Comment by luigidellaquila** [comment]
Hi @sowee15 

the build is platform independent, so no problem.
Releases are published on the website and on the official community group:
https://groups.google.com/forum/#!forum/orient-database

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hotfix releases are minor production releases, in particular, the official stable release is latest hotfix

Thanks

Luigi

**Comment by sowee15** [comment]
Thanks for the info Luigi!

**Comment by sowee15** [comment]
Hi @luigidellaquila,

The reported issue now works fine on the snapshot, thanks you very much! 

However, I tried adding an embeddedmap of date to check if containsvalue and containskey work as expected. Going on from the previous test case:

````
create property concept_dates.dates_map embeddedmap date
update concept_dates set dates_map = { "created": "2015-10-22", "modified": "2015-10-24" }
````

Then those SELECT return no result:

````
select * from concept_dates where dates_map containsvalue "2015-10-22"
select * from concept_dates where dates_map containsvalue "2015-10-22 00:00:00"
select * from concept_dates where dates_map containsvalue "2015-10-24"
select * from concept_dates where dates_map containsvalue "2015-10-24 00:00:00"
````

However this returns 1 result:

````
select * from concept_dates where dates_map containskey "created"
````

I also tried it after creating an index on the field, same result (no result when results were expected).

I tried the same test case but with a map of integer (numbers_map) and everything works fine and it also worked with an index on the field. However, I noticed that creating a predicate like:

````
where numbers_map["created"] = 2
````

does not use the index (I explain-ran the query). I need to do something along the lines of that to optimize the query:

````
where numbers_map containskey "created" and numbers_map containsvalue 2 and numbers_map["created"] = 2
````

That uses the index to filter out the records that contains both the key and the value and then check which one have the value for the particular key. Is this expected?

Again, thanks a lot for your work.

**Comment by luigidellaquila** [comment]
Hi @sowee15 

ok, I'm keeping this open to check consistency on maps.

**Comment by sowee15** [comment]
Hi Luigi,

Were you able to check out the containsmap operator?

Thanks!

**Comment by luigidellaquila** [comment]
Hi @sowee15 

I'm checking it right now

**Comment by luigidellaquila** [comment]
Hi @sowee15 
I just pushed a fix on branch 2.1.x, now I'm porting it on develop branch

**Comment by sowee15** [comment]
Thanks!
